### PR TITLE
Adding Machine Fingerprint to Domain Server permissions

### PR DESCRIPTION
**Pull request by davidkelly**
On windows and mac, we can get a UUID which identifies the computer fairly uniquely.  To spoof it, you would need to recompile interface, or maybe edit your bios (on windows).   Still, a bit harder than spoofing IP or MAC.  It is the bios uuid, which is _usually_ there and unique.  If not, we have a fallback.

For right now, I'm just adding it as another way of identifying a user that isn't logged in.  When you kick a user who is not logged in, we add their IP address and MAC address to the permissions in the domain server.  Now, we add the 'Machine Fingerprint' too, as another identifier.  Future work will make this (and ip/mac address) a bit easier to use.  

Note:  this involves a protocol change.

Test Plan:
There is a section now for Machine Fingerprint in the security settings on the Domain Server settings page.  It works the same as IP or MAC address.  So a logged in user, when kicked, has his/her user name added to that section of the permissions.  However if not logged in, a kicked user has his/her IP, MAC, and Machine Fingerprint all added.  So testing this is same as testing permissions in general, just that you now have an extra parameter.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4d7f3051a7d8f490ae13c1af7ce829aff78461f2 available at: https://gist.github.com/19b361328eeed866cb64013829685cc9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9221-4d7f305.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9221-4d7f305.dmg

**Comment by RyanDowne** [comment]
[integrate fingerprint data](https://trello.com/c/0ksFVo73/31-integrate-fingerprint-data)

File: domain-server/src/DomainGatekeeper.cpp [hunk 1]
```diff
@@ -120,19 +120,21 @@ void DomainGatekeeper::processConnectRequestPacket(QSharedPointer<ReceivedMessag
         nodeData->setPlaceName(nodeConnection.placeName);
 
         qDebug() << "Allowed connection from node" << uuidStringWithoutCurlyBraces(node->getUUID())
-            << "on" << message->getSenderSockAddr() << "with MAC" << nodeConnection.hardwareAddress;
+            << "on" << message->getSenderSockAddr() << "with MAC" << nodeConnection.hardwareAddress
+            << " and machine fingerprint " << nodeConnection.machineFingerprint;
```
Comment (howard-stearns) [inline domain-server/src/DomainGatekeeper.cpp]: I don't think you need the leading/trailing whitespace here and on new line 132. qDebug will add the spaces for you. (Except when it doesn't, and I've never really figure out when/why it does not add the whitespace.)

File: domain-server/src/DomainGatekeeper.h [hunk 2]
```diff
@@ -107,8 +107,8 @@ private slots:
     QSet<QString> _domainOwnerFriends; // keep track of friends of the domain owner
     QSet<QString> _inFlightGroupMembershipsRequests; // keep track of which we've already asked for
 
-    NodePermissions setPermissionsForUser(bool isLocalUser, QString verifiedUsername,
-                                          const QHostAddress& senderAddress, const QString& hardwareAddress);
+    NodePermissions setPermissionsForUser(bool isLocalUser, QString verifiedUsername, const QHostAddress& senderAddress, 
+                                            const QString& hardwareAddress, const QUuid& machineFingerprint);
```
Comment (howard-stearns) [inline domain-server/src/DomainGatekeeper.h]: whitespace

File: libraries/networking/src/FingerprintUtils.cpp [hunk 3]
```diff
@@ -45,6 +50,13 @@ QString FingerprintUtils::getMachineFingerprintString() {
     IWbemLocator *pLoc = NULL;
 
     // initialize com
+    hres = CoInitializeEx(0, COINIT_MULTITHREADED);
+    if (FAILED(hres)) {
+        qDebug() << "Failed to initialize COM library!";
+        return uuidString;
+    }
```
Comment (howard-stearns) [inline libraries/networking/src/FingerprintUtils.cpp]: what's this about? maybe a comment?

File: libraries/networking/CMakeLists.txt [hunk 4]
```diff
@@ -32,7 +33,7 @@ target_link_libraries(${TARGET_NAME} ${OPENSSL_LIBRARIES} ${TBB_LIBRARIES})
 
 # IOKit is needed for getting machine fingerprint
 if (APPLE)
-    target_link_libraries(${TARGET_NAME} ${FRAMEWORK_IOKIT})
+    target_link_libraries(${TARGET_NAME} ${FRAMEWORK_IOKIT} ${CORE_FOUNDATION})
```
Comment (howard-stearns) [inline libraries/networking/CMakeLists.txt]: what is this and the line 20 change for?

Comment (davidkelly) [inline libraries/networking/CMakeLists.txt]: turns out ice-server and vhacd-util both don't link CoreFoundation, but everything else does (in macosx).  I think gustavo doesn't build those, so this worked fine until I built on my mac today.

File: domain-server/src/DomainGatekeeper.cpp (hunk 1 above)
Comment (davidkelly) [inline domain-server/src/DomainGatekeeper.cpp]: hmm ok, cool.

File: libraries/networking/src/FingerprintUtils.cpp (hunk 3 above)
Comment (davidkelly) [inline libraries/networking/src/FingerprintUtils.cpp]: Interface calls CoInitializeEx, but apparently others (agents, etc...) don't.  Now that machine fingerprint is part of the domain connect string, I need to be sure it has been called.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0553dbdcb733298340d273aa8241a4b2cbc72bc3 available at: https://gist.github.com/19b361328eeed866cb64013829685cc9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9221-0553dbd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9221-0553dbd.dmg

**Comment by OmegaHeron** [comment]
QA Pass

Tested with following config
Sandbox/Interface Windows 10
Interface second machine, Ubuntu Linux 16.04.1 (This PR code - of course)

Test 1 - Linux box connecting to windows box AV not logged in
When I kicked AV values were added to domain server for ip address, mac address and machine fingerprint.  Confirmed client could not return until I removed 3 blocking idents.  Also confirmed that logging in client maintained block.  Removed blocking data and client was able to return.

Test 2 - Same as above except client from Linux machine was logged in with HF account.  Results same as above.  I sort of expected an entry to be made in User block with logged in name... overkill perhaps.

This works as advertised.

File: domain-server/resources/describe-settings.json [hunk 5]
```diff
@@ -900,6 +900,78 @@
...
+            {
+              "name": "id_can_connect_past_max_capacity",
+              "label": "Ignore Max Capacity",
+              "type": "checkbox",
+              "editable": true,
+              "default": false
+            },
+            {
+              "name": "id_can_kick",
+              "label": "Kick Users",
```
Comment (zfox23) [inline domain-server/resources/describe-settings.json]: Is now a good time to start changing "Kick" to "Ban" ?

**Pull request merged**